servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/TcpFastOpenTest.java

```diff
@@ -0,0 +1,109 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.internal.ServiceTalkTestTimeout;
+import io.servicetalk.http.api.BlockingHttpClient;
+import io.servicetalk.http.api.HttpResponseStatus;
+import io.servicetalk.http.api.HttpServerBuilder;
+import io.servicetalk.http.api.SingleAddressHttpClientBuilder;
+import io.servicetalk.transport.api.HostAndPort;
+import io.servicetalk.transport.api.ServerContext;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.net.InetSocketAddress;
+import java.net.SocketOption;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static io.servicetalk.transport.api.ServiceTalkSocketOptions.TCP_FASTOPEN_BACKLOG;
+import static io.servicetalk.transport.api.ServiceTalkSocketOptions.TCP_FASTOPEN_CONNECT;
+import static io.servicetalk.transport.netty.internal.AddressUtils.localAddress;
+import static io.servicetalk.transport.netty.internal.AddressUtils.serverHostAndPort;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonMap;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Parameterized.class)
+public class TcpFastOpenTest {
+    @Rule
+    public final Timeout timeout = new ServiceTalkTestTimeout();
+    @SuppressWarnings("rawtypes")
+    private final Map<SocketOption, Object> serverListenOptions;
+    @SuppressWarnings("rawtypes")
+    private final Map<SocketOption, Object> clientOptions;
+
+    public TcpFastOpenTest(
+            @SuppressWarnings("rawtypes") final Map<SocketOption, Object> serverListenOptions,
+            @SuppressWarnings("rawtypes") final Map<SocketOption, Object> clientOptions) {
+        this.serverListenOptions = serverListenOptions;
+        this.clientOptions = clientOptions;
+    }
+
+    @Parameterized.Parameters(name = "{index}: server opts={0} client opts={1}")
+    public static Collection<Object[]> sslProviders() {
+        return asList(
+                new Object[]{emptyMap(), emptyMap()},
+                new Object[]{emptyMap(), clientTcpFastOpenOptions()},
+                new Object[]{serverTcpFastOpenOptions(), emptyMap()},
+                new Object[]{serverTcpFastOpenOptions(), clientTcpFastOpenOptions()}
+        );
+    }
+
+    @SuppressWarnings("rawtypes")
+    static Map<SocketOption, Object> clientTcpFastOpenOptions() {
+        return singletonMap(TCP_FASTOPEN_CONNECT, true);
+    }
+
+    @SuppressWarnings("rawtypes")
+    static Map<SocketOption, Object> serverTcpFastOpenOptions() {
+        return singletonMap(TCP_FASTOPEN_BACKLOG, 1);
+    }
+
+    @Test
+    public void requestSucceedsEvenIfTcpFastOpenNotEnabledOrSupported() throws Exception {
```

Comment:
Btw, this test proves that it's safe to have this option always enabled. WDYT about making it on by default (opt-out)?

Reply:
Netty logs warnings if you enable an option that isn't supported, and also if it is supported users may not want to have it always enabled (e.g. idempotent, extra state tracking, etc.).

Reply:
Thanks for the explanation. What if we move it down to the transport, where we have access to `Epoll.isAvailable()` check (or `Channel` type) and know that the `Channel` is secure? For example, `SslClientChannelInitializer` might be a good place, if we pass the user-defined options there as well (to check that user did not explicitly set this option).

Reply:
Interesting idea. I propose we proceed by adding the feature as opt-in first, and after we get some data we can consider enabling by default in a followup PR.

Reply:
Sounds good to me!

Reply:
Reading around privacy concerns with TFO [1], makes me think that it's probably better left as opt-in. 
1. https://arxiv.org/pdf/1905.03518.pdf